pkgs/mkService.nix

```diff
@@ -77,16 +77,25 @@ let
           RuntimeDirectory = systemdRequiredPaths runtimeDirs;
           RuntimeDirectoryMode = runtimeDirsMode;
        };
+
   execStartPre = optionalScript "${name}-prestart" (lib.concatStrings [
     (lib.optionalString user.create ''
       # Setup user
       if ! ${stdenv.glibc.bin}/bin/getent passwd ${user.name} > /dev/null; then
         ${remoteShadow}/bin/useradd --system --user-group ${user.name} --home ${if user.home == null then "/var/empty" else user.home}
       fi
+
+      # Ensure user groups initialised (if any) and then add user to groups (if any)
+      ${(lib.concatStrings (map (group: "${remoteShadow}/bin/groupadd -f ${group}\n") user.groups))}
+      ${(lib.concatStrings (map (group: "${remoteShadow}/bin/usermod -a -G ${group} ${user.name}\n") user.groups))}
+
+      # Create user home
```

Comment:
I think `# Setup user` covers these, and in a lot of cases these comments will be followed by empty space.